google/cloud/bigtable/internal/completion_queue_impl.h

```diff
@@ -126,13 +146,98 @@ class AsyncUnaryRpcFunctor : public AsyncOperation {
+              CheckUnaryStreamRpcDataCallback<DataFunctor, Response>::value,
+              int>::type = 0,
+          typename std::enable_if<CheckUnaryStreamRpcFinishedCallback<
+                                      FinishedFunctor, Response>::value,
+                                  int>::type = 0>
+class AsyncUnaryStreamRpcFunctor : public AsyncOperation {
+ public:
+  explicit AsyncUnaryStreamRpcFunctor(DataFunctor&& data_functor,
+                                      FinishedFunctor&& finished_functor)
+      : state_(CREATING),
+        data_functor_(std::forward<DataFunctor>(data_functor)),
+        finished_functor_(std::forward<FinishedFunctor>(finished_functor)) {}
+
+  /// Make the RPC request and prepare the response callback.
+  template <typename Client, typename MemberFunction>
+  void Set(Client& client, MemberFunction Client::*call,
+           std::unique_ptr<grpc::ClientContext> context, Request const& request,
+           grpc::CompletionQueue* cq, void* tag) {
+    tag_ = tag;
+    context_ = std::move(context);
+    response_reader_ = (client.*call)(context_.get(), request, cq, tag);
+  }
+
+  void Cancel() override { context_->TryCancel(); }
+
+ private:
+  enum State { CREATING = 4525, PROCESSING = 21345, FINISHING = 82553 };
+  bool Notify(CompletionQueue& cq, Disposition d) override {
+    // Due to issue #1308, argument d doesn't mean what the state names suggest.
+    // If d == AsynOperation::CANCELLED, it might as well be a broken stream.
+
+    std::unique_lock<std::mutex> lk(mu_);
+
+    switch (state_) {
+      case CREATING:
+        if (d == AsyncOperation::COMPLETED) {
+          response_reader_->Read(&response_, tag_);
+          state_ = PROCESSING;
+        } else {
+          response_reader_->Finish(&status_, tag_);
+          state_ = FINISHING;
+        }
+        return false;
+      case PROCESSING:
+        // Optimization is possible here. We could kick off
+        // response_reader_->Read() before calling data_functor_. If we did,
+        // we'd potentially utilize more threads executing the completion queue.
+        // However, we'd have to either add extra code to make sure that the
+        // callbacks are run in the same order as they arrived or introduce a
+        // mode in which they can be reordered.
+        if (d == AsyncOperation::COMPLETED) {
+          Response received;
+          response_.Swap(&received);
+          lk.unlock();
+          data_functor_(cq, *context_, received);
+          lk.lock();
+          response_reader_->Read(&response_, tag_);
+        } else {
+          response_reader_->Finish(&status_, tag_);
+          state_ = FINISHING;
+        }
+        return false;
+      case FINISHING:
+        finished_functor_(cq, *context_, status_);
+        return true;
+    }
+    google::cloud::internal::RaiseRuntimeError(
+        "unexpected state in AsyncUnaryRpcFunctor: " + std::to_string(state_));
+  }
+
+  std::mutex mu_;
```

Comment:
TBH, it's an accidental leftover from the code I had to run on multiple threads simultaneously, but before I realized that I realized that it could reorder the callbacks.

Having put more thought into it, I think there are legitimate reasons for it to be there.

Reply:
Ack thanks!